src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingTimelineService.kt

```diff
@@ -27,54 +29,81 @@ class BookingTimelineService(
     val sortedSentences = booking.getAllExtractableSentences().sortedBy { it.sentencedAt }
     val firstSentence = sortedSentences[0]
     var sentenceRange = firstSentence.getRangeOfSentenceBeforeAwardedDays()
     val totalAda = firstSentence.sentenceCalculation.calculatedTotalAwardedDays
     var previousSentence = firstSentence
 
-    var daysAwardedServed = 0
-    var deductionsNoLongerApplicable = 0
+    // Which sentences are in the same "Group". A "Group" of sentences are sentences that are concurrent to each other,
+    // and there sentenceAt dates overlap with the other release date. i.e. there is no release inbetween them
+    // Whenever a release happens a new group is started.
+    val sentencesInGroup: MutableList<ExtractableSentence> = mutableListOf()
+    var lastReleaseDateReached: LocalDate? = null
 
     sortedSentences.forEach {
+      it.sentenceCalculation.adjustmentsBefore = sentenceRange.end
+      it.sentenceCalculation.adjustmentsAfter = lastReleaseDateReached
       val itRange = it.getRangeOfSentenceBeforeAwardedDays()
       if (sentenceRange.isConnected(itRange)) {
         if (itRange.end.isAfter(sentenceRange.end)) {
           sentenceRange = LocalDateRange.of(sentenceRange.start, itRange.end)
         }
+        sentencesInGroup.add(it)
       } else {
         // There is gap here.
-
         // 1. Check if there have been any ADAs served
         var daysBetween = ChronoUnit.DAYS.between(sentenceRange.end, it.sentencedAt)
         val daysAdaServed = min(daysBetween - 1, totalAda.toLong())
-        daysAwardedServed += daysAdaServed.toInt()
-        // Update range to include ada's served.
-        sentenceRange = LocalDateRange.of(sentenceRange.start, previousSentence.sentenceCalculation.adjustedReleaseDate)
+        booking.adjustments.addAdjustment(
+          AdjustmentType.ADDITIONAL_DAYS_SERVED,
+          Adjustment(
+            numberOfDays = daysAdaServed.toInt(),
+            appliesToSentencesFrom = it.sentencedAt
+          )
+        )
 
-        daysBetween = ChronoUnit.DAYS.between(sentenceRange.end, it.sentencedAt)
-        if (daysBetween > 0) {
+        daysBetween = ChronoUnit.DAYS.between(sentenceRange.end.plusDays(daysAdaServed), it.sentencedAt)
+        if (daysBetween <= 1) {
+          // The gap has been filled by served adas.
+          sentenceRange = LocalDateRange.of(sentenceRange.start, itRange.end)
+          sentencesInGroup.add(it)
+        } else {
           // There is still a gap
 
           // 2. A release date has occurred but there are more sentences on the booking, therefore previous deductions
           // should be wiped.
-          deductionsNoLongerApplicable += booking.getOrZero(
-            AdjustmentType.REMAND, AdjustmentType.TAGGED_BAIL,
-            adjustmentsFrom = previousSentence.sentenceCalculation.adjustedReleaseDate
-          )
+          lastReleaseDateReached = previousSentence.sentenceCalculation.adjustedReleaseDate
```

Comment:
Minor one - the word `last` can be misconstrued here, maybe call the var `previousReleaseDateReached`

Reply:
Done